### PR TITLE
Use optional-lite wherever makes sense

### PR DESCRIPTION
**Pull request by Minoru**
The goal here is to streamline some of the interfaces and move our C++ a bit closer to Rust. Now we can refactor to `optional` on the C++ side, and move that code to Rust with its `Option`. This is easier, more pleasant, and less error-prone than dealing with pairs of a `bool` and a value.

Refactoring `Matchable` interface required tests for its implementations, which in turn required `EnvVar` to run `setlocale()`. I also ran into a couple bugs while testing, so fixed those as well… That's why there are so many "preparatory" commits here. Happy to split into multiple PRs if it's too hard to review.

Reviews from everyone are welcome. Will merge in three days.

### CODE REVIEW

File: test/rssfeed.cpp [hunk 1]
```diff
@@ -393,3 +393,164 @@ TEST_CASE(
...
 }
+
+TEST_CASE("RssFeed contains a number of matchable attributes", "[RssFeed]")
+{
+	ConfigContainer cfg;
+	Cache rsscache(":memory:", &cfg);
+	RssFeed f(&rsscache);
+
+	SECTION("`feedtitle`, containing feed's title") {
+		const auto title = std::string("Example feed");
```
[the remark this answers is not in the document]
Comment (Minoru) [reply test/rssfeed.cpp:407]: No advantage, just a more fashionable way. Modern C++ pushes the "almost always auto" attitude, which basically mandates me to write `const auto title` here, and then I have to put the type on the RHS if I want to specify it. I still find this approach a bit unusual, so I try it from time to time to see if it sticks. Still doesn't :)

File: test/rssfeed.cpp [hunk 2]
```diff
@@ -393,3 +393,164 @@ TEST_CASE(
...
+		f.set_title(title);
+
+		const auto attr = "feedtitle";
+		REQUIRE(f.has_attribute(attr));
+		REQUIRE(f.get_attribute(attr) == title);
```
[the remark this answers is not in the document]
Comment (Minoru) [reply test/rssfeed.cpp]: Oh, you're right, it does! When I wrote this test, I didn't look past [`RssFeed::get_attribute()`](https://github.com/newsboat/newsboat/blob/4edcbddc4f0d27c4df6a736df45b6edf08baf049/src/rssfeed.cpp#L146), and only added charset tests for attributes for which `get_attribute` calls `utils::ut8_to_locale()`. But I now see that [`RssFeed::title()` also calls `utils::utf8_to_locale()`](https://github.com/newsboat/newsboat/blob/4edcbddc4f0d27c4df6a736df45b6edf08baf049/src/rssfeed.cpp#L99).

I'll update the tests.

File: test/test-helpers/envvar.cpp [hunk 3]
```diff
@@ -303,6 +320,20 @@ TEST_CASE("EnvVar::unset() runs a function (set by on_change()) after changing "
 		REQUIRE(counter == 2);
 	}
 
+	SECTION("Function is passed `nullopt`") {
```
[the remark this answers is not in the document]
Comment (Minoru) [reply test/test-helpers/envvar.cpp:323]: But the "unset" part is already mentioned in the name of the test case. When an assertion here fails, you'll see something like this:

```
-------------------------------------------------------------------------------
EnvVar::unset() runs a function (set by on_change()) after changing the
environment variable
  Function is passed `nullopt`
-------------------------------------------------------------------------------
test/test-helpers/envvar.cpp:323
...............................................................................

test/test-helpers/envvar.cpp:324: FAILED:
explicitly with message:
  example test failure
```

which I believe is plenty of info to understand what's going on. Do you disagree?

File: src/matcher.cpp [hunk 4]
```diff
@@ -83,11 +82,15 @@ bool Matcher::matches(Matchable* item)
 
 bool Matcher::matchop_lt(expression* e, Matchable* item)
 {
-	if (!item->has_attribute(e->name))
+	const auto attr = item->attribute_value(e->name);
+
+	if (!attr.has_value()) {
 		throw MatcherException(
 			MatcherException::Type::ATTRIB_UNAVAIL, e->name);
+	}
```
[the remark this answers is not in the document]
Comment (Minoru) [reply src/matcher.cpp:90]: I feel this PR is big enough already, but okay, it's close enough to the boyscout rule that I can do this :)

File: test/test-helpers/envvar.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/test-helpers/envvar.cpp:323]: Oh, I guess I somehow looked at the wrong `TEST_CASE` message.

Looks good!

File: src/matcher.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/matcher.cpp:92]: Thanks!

**Comment by Minoru** [comment]
How do new changes look, @dennisschagt?

**Comment by Minoru** [comment]
Rebased to apply fixups, the result is the same: https://github.com/newsboat/newsboat/compare/8dd14368a78953271fb03e9c24e1597e7e3a599f..d89b3a3dcdcca9dc7a8020cb34ca1f3f15771dc9 Will merge once all the checks are green.

Thanks for the review, @dennisschagt1

**Pull request merged**